### ISSUE TITLE
Dungeon: Implementiere ein einfaches Spiel mit 4..5 Leveln 

### ISSUE THREAD

**Issue by cagix** [Programmiermethoden/Dungeon#1355]
Für den Einsatz in Prog2 brauchen wir ein einfaches Spiel, welches ca. 4..5 Level hat. Jedes Level sollte klar erkennbare Strukturen haben (also zumindest optisch klar unterscheidbar sein). Auf jedem Level sollten andere Monster agieren, diese sollten zunehmend schwieriger zu besiegen sein. Es soll Quests geben, die zunächst nur unterschiedlich sein sollen (etwa das Besiegen von N Monstern, oder das Aufsammeln von N Tränken o.ä.). Das Tor zum nächsten Level darf erst aufgehen, wenn alle Monster und Aufgaben erledigt sind. 

Das Spiel soll so organisiert werden, dass die Studierenden darin Modifikationen vornehmen können und dabei die ECS-Architektur kennen lernen (und Spielspass haben). Die Modifikationsmöglichkeiten sollen von einfach ("ändere Eigenschaften von Monster X") bis komplex ("ändere das Verhalten von Monster X" oder "ändere die Quest, so dass ...") reichen.

Es können dazu aus dem Projekt `Game` die Klassen in `core` und aus dem Projekt `Dungeon` die Klassen in `contrib` verwendet werden.

**Comment by AMatutat** [comment]
> Für die Steuerung sollten nicht nur die klassischen WASD konfiguriert werden, sondern zusätzlich auch die Cursor-Tasten. Zusätzlich wäre es nett, wenn man per Mausklick steuern könnte (@AMatutat ist das überhaupt implementiert?).

Du kannst mit der (jetzt lass mich nicht lügen) Linken Maustaste Interagieren (z.B Kisten aufmachen) und mit der rechten Maustaste feuerspucken. 

Bewegen per Maus geht nicht. Das ist auch ein etwas "komplexeres" Problem (hat einige Edge-Cases).

**Comment by cagix** [comment]
ja, und feuern geht auch in richtung maus, wenn ich mich richtig erinnere?

**Comment by AMatutat** [comment]
> ja, und feuern geht auch in richtung maus, wenn ich mich richtig erinnere?

so hab ich das auch im kopf

**Comment by cagix** [comment]
> Bewegen per Maus geht nicht. Das ist auch ein etwas "komplexeres" Problem (hat einige Edge-Cases).

Das klingt nach einer Aufgabe im Praktikum für die Studis? Offenbar kann man die Position des Mauszeigers bereits im Spiel abfragen. Dann müsste man nur noch einen Vektor von der aktuellen Position zur Mausposition basteln und sich entsprechend der Velocity in diese Richtung bewegen. Allerdings nur, wenn man irgendwo hin klickt. 

Man könnte dann noch überlegen, ob das dann immer nur eine bestimmte maximale Distanz sein soll oder ob man mit einem Wegesuch-Algorithmus einen Weg findet und dann entsprechend dorthin läuft. Im Original wird der Weg dabei durch die bereits erkundeten Gebiete gefunden (haben wir nicht), und die Bewegung geht nur so lange, bis irgendwas unterbricht (Monster, Hunger, ...).

**Comment by cagix** [comment]
@Denniso3 Ich hatte heute ein Gespräch mit unseren Kollegen vom DigikoS-Projekt. Die haben eine Art 2D-Map von Bielefeld (stark vereinfacht) mit ein paar Gebäuden als Grafik und ein paar Monstern (Grafiken in verschiedenen Zuständen: 100% Leben, 50% Schaden, 0% Leben). Damit wollen wir im ILIAS eine Art Lernspiel aufbauen, wo wöchentlich ein neues Monster eine bestimmte Landmark bedroht und durch das Lösen eines Quizzes besiegt werden kann. Dabei gehen die Punkte aller Teilnehmer in den Score, und wenn eine bestimmte Menge an Punkten erreicht ist, dann ist das Monster besiegt.

Im Gespräch kam die Idee auf, ob man das nicht auf das von Dir entwickelte Spiel übertragen kann. Im Prinzip wäre das ein Level mit "Bielefeld-Layout", also man müsste die Kacheln einfach mit den Grafiken aus dem DigikoS-Projekt belegen und die Grafiken der Monster in die richtigen Animation-Ordner packen. Dann könnte man das auch als eine Ebene im Spiel benutzen und hier für das Praktikum noch Programmieraufgaben einbauen.

**Comment by Denniso3** [comment]
Ja das klingt interessant, die Idee könnte man wirklich übernehmen. 

> Dabei gehen die Punkte aller Teilnehmer in den Score, und wenn eine bestimmte Menge an Punkten erreicht ist, dann ist das Monster besiegt. 
>  

Soll das mit "allen Teilnehmer" auch irgendwie übertragen werden oder soll der eine "Studi" einen gewissen Score erreichen und das Monster ist dann besiegt? Oder soll es hauptsächlich nur um das neue "Bielefeld-Layout" gehen + Monster bedroht Landmark?

**Comment by cagix** [comment]
> Ja das klingt interessant, die Idee könnte man wirklich übernehmen.
> 
> > Dabei gehen die Punkte aller Teilnehmer in den Score, und wenn eine bestimmte Menge an Punkten erreicht ist, dann ist das Monster besiegt.
> 
> Soll das mit "allen Teilnehmer" auch irgendwie übertragen werden oder soll der eine "Studi" einen gewissen Score erreichen und das Monster ist dann besiegt? Oder soll es hauptsächlich nur um das neue "Bielefeld-Layout" gehen + Monster bedroht Landmark?

Das ist vor allem für ILIAS gedacht - dort haben alle den selben View und man (aka ich) muss dann die Ergebnisse aus den Quizzes manuell zusammenführen und in den View übertragen.

Hier für das Spiel kann man das so gar nicht umsetzen - hier programmiert und spielt ja jedes Team alleine. Aber vielleicht kann man eines der vier...fünf Level entsprechend "tapezieren"?
